File: src/main/python/g0001_0100/s0007_reverse_integer/Solution.py
# #Medium #Top_Interview_Questions #Math #Udemy_Integers
# #2024_06_03_Time_28_ms_(94.06%)_Space_16.5_MB_(93.17%)

class Solution:
    def reverse(self, x: int) -> int:
        rev = 0
        sign = -1 if x < 0 else 1
        x = abs(x)

        while x != 0:
            rev = rev * 10 + x % 10
            x //= 10

        rev *= sign

        if rev > 2**31 - 1 or rev < -2**31:
            return 0

        return rev

File: src/main/dart/g0001_0100/s0007_reverse_integer/Solution.dart
// #Medium #Top_Interview_Questions #Math #Udemy_Integers
// #2024_10_03_Time_347_ms_(96.43%)_Space_153.6_MB_(5.61%)

class Solution {
  int reverse(int x) {
    int number = x.abs();
    int reverseNumber = 0;
    final int maxValue = 0x7fffffff;
    while (number != 0) {
      reverseNumber = reverseNumber * 10 + number % 10;
      number = number ~/ 10;
    }
    if (reverseNumber > maxValue) {
      return 0;
    }
    return x > 0 ? reverseNumber : -reverseNumber;
  }
}
